### PR TITLE
STORM-2497: Let Supervisor enforce memory and add in support for shared memory regions

### PR DESCRIPTION
**Pull request by revans2**
This is based off of #2112 so that the formatting is simpler.

### CODE REVIEW

**Comment by jerrypeng** [comment]
cool feature

**Comment by revans2** [comment]
@kishorvpatil Could you look again.  I had to rebase becase of a minor conflict in the DefaultResourceAwareStrategy.  It ended up resulting in only some comment changes to this code, but I wanted to be sure you took a look.

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java [hunk 1]
```diff
@@ -728,8 +758,145 @@ protected String javaCmd(String cmd) {
...
+        hardMemoryLimitOver = this.hardMemoryLimitOver * getTotalWorkersForThisTopology();
+        typeOfCheck = "TOPOLOGY " + _topologyId;
+      } else {
+        usageMb = getMemoryUsageMb();
+        memoryLimitMb = this.memoryLimitMB;
+        hardMemoryLimitOver = this.hardMemoryLimitOver;
+        typeOfCheck = "WORKER " + _workerId;
+      }
+      LOG.debug(
+          "Enforcing memory usage for {} with usgae of {} out of {} total and a hard limit of {}",
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java]: usgae -> usage

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java [hunk 2]
```diff
@@ -728,8 +758,145 @@ protected String javaCmd(String cmd) {
...
+          typeOfCheck,
+          usageMb,
+          memoryLimitMb,
+          hardMemoryLimitOver);
+
+      if (usageMb <= 0) {
+        //Looks like usage might now be supported
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java]: now -> not(?)

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java [hunk 3]
```diff
@@ -728,8 +758,145 @@ protected String javaCmd(String cmd) {
...
+        return false;
+      }
+      long hardLimitMb =
+          memoryLimitMb
+              + Math.max(
+                  (long) (memoryLimitMb * (hardMemoryLimitMultiplier - 1.0)), hardMemoryLimitOver);
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java]: Nit: I think this would be easier to read as max(memoryLimit*hardMultiplier, memoryLimit + hardLimitOver), rather than adding first and then adjusting the multiplier.

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java [hunk 4]
```diff
@@ -528,28 +577,108 @@ public void cleanUpForRestart() throws IOException {
...
         _workerId = null;
     }
-    
+
+    /**
+     * Check if the container is over its memory limit AND needs to be killed. This does not necessarily mean
+     * that it just went over the limit.
+     * @throws IOException on any error
+     */
+    public boolean isMemoryLimitViolated(LocalAssignment withUpdatedLimits) throws IOException {
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java]: This doesn't seem like it can throw IOException

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java [hunk 5]
```diff
@@ -559,9 +688,17 @@ public void cleanUpForRestart() throws IOException {
...
+     * Get the id of the container or null if there is no worker id right now.
      */
     public String getWorkerId() {
         return _workerId;
     }
+
+    /**
+     * Get the topology id this is a part of.
+     */
+    public String getTopoogyId() {
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java]: topoogy -> topology

File: storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.storm.Config;
+import org.apache.storm.DaemonConfig;
+import org.apache.storm.generated.StormTopology;
+import org.apache.storm.testing.TestWordSpout;
+import org.apache.storm.topology.TopologyBuilder;
+import org.junit.Test;
+
+public class NimbusTest {
+    @Test(expected=IllegalArgumentException.class)
```
Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java]: Nit: Could reduce the scope of this assertion more local with the ExpectedException rule

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/TestResourceAwareScheduler.java [hunk 7]
```diff
@@ -46,53 +44,43 @@
...
     public static void initConf() {
-        defaultTopologyConf.put(Config.STORM_NETWORK_TOPOGRAPHY_PLUGIN, "org.apache.storm.networktopography.DefaultRackDNSToSwitchMapping");
-        defaultTopologyConf.put(DaemonConfig.RESOURCE_AWARE_SCHEDULER_EVICTION_STRATEGY, org.apache.storm.scheduler.resource.strategies.eviction.DefaultEvictionStrategy.class.getName());
-        defaultTopologyConf.put(DaemonConfig.RESOURCE_AWARE_SCHEDULER_PRIORITY_STRATEGY, org.apache.storm.scheduler.resource.strategies.priority.DefaultSchedulingPriorityStrategy.class.getName());
-
-        defaultTopologyConf.put(Config.TOPOLOGY_SCHEDULER_STRATEGY, org.apache.storm.scheduler.resource.strategies.scheduling.DefaultResourceAwareStrategy.class.getName());
-        defaultTopologyConf.put(Config.TOPOLOGY_COMPONENT_CPU_PCORE_PERCENT, 10.0);
-        defaultTopologyConf.put(Config.TOPOLOGY_COMPONENT_RESOURCES_ONHEAP_MEMORY_MB, 128.0);
-        defaultTopologyConf.put(Config.TOPOLOGY_COMPONENT_RESOURCES_OFFHEAP_MEMORY_MB, 0.0);
+      //TODO clean this up some more
```
Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/TestResourceAwareScheduler.java]: Todo

File: storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java [hunk 8]
```diff
@@ -144,8 +148,19 @@ public Void call() throws Exception {
...
+                        //We need to create a directory for shared memory to write to (we should not encourage this though)
+                        Path path = sharedMemoryDirTmpLocation.toPath();
+                        Files.createDirectories(path);
+                    }
                     _fsOps.moveDirectoryPreferAtomic(tr, _stormRoot);
-                    _fsOps.setupStormCodeDir(ConfigUtils.readSupervisorStormConf(_conf, _topologyId), _stormRoot);
+                    Map<String, Object> topoConf = ConfigUtils.readSupervisorStormConf(_conf, _topologyId);
+                    _fsOps.setupStormCodeDir(topoConf, _stormRoot);
+                    if (_assignment.is_has_node_shared_memory()) {
+                        File sharedMemoryDir = new File(_stormRoot, "shared_by_topology");
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java]: I might have missed it somewhere, but how does the topology code access the off heap shared memory?

File: storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java [hunk 9]
```diff
@@ -151,22 +161,40 @@ public void reserveResourcesForWorker(String workerId, Map<String, Number> resou
...
-                memCore.setPhysicalUsageLimit(Long.valueOf(totalMem.longValue() * 1024 * 1024));
-            } catch (IOException e) {
-                throw new RuntimeException("Cannot set memory.limit_in_bytes! Exception: ", e);
+        if ((boolean) this.conf.get(DaemonConfig.STORM_CGROUP_MEMORY_ENFORCEMENT_ENABLE)) {
+            if (totalMem != null) {
+                int cgroupMem =
+                    (int)
+                        (Math.ceil(
+                            ObjectReader.getDouble(
+                                this.conf.get(DaemonConfig.STORM_CGROUP_MEMORY_LIMIT_TOLERANCE_MARGIN_MB),
```
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java]: I'm wondering if having defaults here is necessary? Isn't defaults.yaml always expected to be present?

File: storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java [hunk 10]
```diff
@@ -19,36 +19,82 @@
...
-        if (executorToSlots != null) {
-            this.executorToSlot.putAll(executorToSlots);
+    private final Map<ExecutorDetails, WorkerSlot> executorToSlot = new HashMap<>();
+    private final Map<WorkerSlot, WorkerResources> resources = new HashMap<>();
+    private final Map<String, Double> totalSharedOffHeap = new HashMap<>();
+
+    public SchedulerAssignmentImpl(String topologyId, Map<ExecutorDetails, WorkerSlot> executorToSlot,
+            Map<WorkerSlot, WorkerResources> resources, Map<String, Double> totalSharedOffHeap) {
+        this.topologyId = topologyId;       
+        if (executorToSlot != null) {
```
Comment (revans2) [inline storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java]: I prefer to be very lenient with inputs if something can be null I don't want to force others to never use null.

File: storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java [hunk 11]
```diff
@@ -141,4 +222,42 @@ public String getTopologyId() {
...
+        totalSharedOffHeap.put(node, value);
+    }
+    
+    @Override
+    public Map<String, Double> getTotalSharedOffHeapMemory() {
+        return totalSharedOffHeap;
+    }
+
+    /**
+     * Update the resources for this assignment (This should go aware when the RAS-MT bridge goes away
```
Comment (revans2) [inline storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java]: Sorry code I forgot to remove that is not needed here.

File: storm-client/src/jvm/org/apache/storm/topology/BaseConfigurationDeclarer.java [hunk 12]
```diff
@@ -52,18 +52,20 @@ public T setNumTasks(Number val) {
         return addConfiguration(Config.TOPOLOGY_TASKS, val);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T setMemoryLoad(Number onHeap) {
         if (onHeap != null) {
```
Comment (revans2) [inline storm-client/src/jvm/org/apache/storm/topology/BaseConfigurationDeclarer.java]: I agree but that is code that came before, and if we want to make that change I would prefer to do it on a separate pull request, just so a breaking change does not sneak in as part of review comments.

Comment (revans2) [inline storm-client/src/jvm/org/apache/storm/topology/BaseConfigurationDeclarer.java]: I filed https://issues.apache.org/jira/browse/STORM-2545 for it

File: storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java [hunk 13]
```diff
@@ -29,8 +33,8 @@
  */
 public class DefaultResourceDeclarer<T extends DefaultResourceDeclarer> implements ResourceDeclarer<T>, ITridentResource {
 
-    private Map<String, Number> resources = new HashMap<>();
-    private static Map<String, Object> conf = Utils.readStormConfig();
+    private final transient Map<String, Number> resources = new HashMap<>();
```
Comment (revans2) [inline storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java]: In trident Node inherits from DefaultResourceDeclarer, and Node is serialized out (repeatedly for different bolts and spouts in the topology).  These are never used after the topology is created so this reduces the serialized size of a trident topology.

I'll add a comment in the code to make it clean why.

File: storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java (hunk 9 above)
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java]: Except in some unit tests.

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java (hunk 4 above)
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java]: Yes but it is overridden by other Container Implementations and they may want to throw an exception.

File: storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java (hunk 8 above)
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java]: That is up to the topology.  It could be JNI code or even something that a ShellBolt or ShellSpout does.  The point is to provide a place for them to store the data if they need it.  We have a group doing this in /tmp right now and causing all kinds of issues if things crash/get rescheduled leaking things.

File: storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java (hunk 10 above)
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java]: Ok. I'm the opposite way, I prefer avoiding null checks. But this is fine :)

File: storm-client/src/jvm/org/apache/storm/topology/BaseConfigurationDeclarer.java (hunk 12 above)
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/topology/BaseConfigurationDeclarer.java]: Thanks

File: storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java (hunk 9 above)
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java]: Makes sense

File: storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java (hunk 8 above)
Comment (srdo) [inline storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java]: Okay. I was just thinking it might be helpful if the topology could get the directory path (or _stormRoot), so users don't have to figure out how to path to this directory themselves.

Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java]: I can add it to the documentation.  But I didn't spend a lot of time documenting it on purpose.  shared memory off heap is hard to get right and I would rather discourage people from using it.

**Comment by revans2** [comment]
I think I addressed all of the review comments

File: storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignment.java [hunk 14]
```diff
@@ -21,41 +21,52 @@
...
      * @param executor
-     * @return
+     * @return true if it is assigned else false
      */
     public boolean isExecutorAssigned(ExecutorDetails executor);
     
     /**
-     * get the topology-id this assignment is for.
-     * @return
+     * Return the ID of the topolgoy.
```
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignment.java]: topolgoy

File: storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java [hunk 15]
```diff
@@ -29,8 +34,12 @@
  */
 public class DefaultResourceDeclarer<T extends DefaultResourceDeclarer> implements ResourceDeclarer<T>, ITridentResource {
 
-    private Map<String, Number> resources = new HashMap<>();
-    private static Map<String, Object> conf = Utils.readStormConfig();
+    //@{link org.apache.storm.trident.planner.Node} and several other tirdent classes inherit from DefaultResourceDeclarer
```
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java]: tirdent

File: storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java [hunk 16]
```diff
@@ -29,8 +34,12 @@
...
+    // These classes are serialized out as part of the bolts and spouts of a topology, often for each bolt/spout in the topology.
+    // The following are marked as transiant because they are never used after the topology is created so keeping them around just wasts
```
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java]: wasts

File: storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java [hunk 17]
```diff
@@ -24,8 +24,10 @@
 import org.apache.storm.topology.TopologyBuilder;
 import org.junit.Test;
 
+import static org.junit.Assert.*;
```
Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java]: I think checkstyle bans this

File: storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java [hunk 18]
```diff
@@ -44,6 +46,11 @@ public void testMemoryLoadLargerThanMaxHeapSize() throws Exception {
         config1.put(Config.TOPOLOGY_SUBMITTER_USER, "zhuo");
         config1.put(Config.TOPOLOGY_WORKER_MAX_HEAP_SIZE_MB, 128.0);
         config1.put(Config.TOPOLOGY_COMPONENT_RESOURCES_ONHEAP_MEMORY_MB, 129.0);
-        Nimbus.validateTopologyWorkerMaxHeapSizeConfigs(config1, stormTopology1);
+        try {
+            Nimbus.validateTopologyWorkerMaxHeapSizeConfigs(config1, stormTopology1);
+            fail("Expected exeption not thrown");
```
Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java]: exeption

File: storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignment.java (hunk 14 above)
Comment (revans2) [inline storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignment.java]: I even typo my fixes.

File: storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java (hunk 17 above)
Comment (revans2) [inline storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java]: Checkstyle does not check tests for some reason, but if you care about it I can change it.

Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java]: No, don't worry about it. Hadn't noticed this was in a test

**Comment by revans2** [comment]
@srdo I think I got the rest of them

@kishorvpatil could you take a look at let me know if the rework is good for you?

**Comment by revans2** [comment]
@srdo @kishorvpatil @jerrypeng 

I just rebased this on the latest master and added in a few caches in some of the internal data structures to help with performance issues we had seen in production.  Please take a look again, I really would like to get this in sooner rather then later as we have some other work we want to do on the RAS scheduler to add in generic resource support (we have a goal to officially support GPUs)

**Comment by srdo** [comment]
+1, though I'll admit to just skimming it this time. It would be easier to tell what changed if the non-merge-conflict-resolution changes were in separate commits.

It might make sense to drop a line to @roshannaik before merging, I think he asked on the mailing list to be made aware of changes to storm-client due to https://github.com/apache/storm/pull/2241
